### ISSUE TITLE
"Binary file has changed" shows for normal text file

### ISSUE THREAD

**Issue by Morilli** [desktop/desktop#7857]
On my repository, one of my untracked files is a file containing debug outputs of the program I was running. While these debug outputs are plain text, GitHub Desktop fails to identify the file as text file and only shows
```
This binary file has changed.
Open file in external program.
```
when selecting that file in the ``Changes`` tab instead of the text that the file contains. 
I believe this is caused by the specific encoding used by that text file (I didn't specifically chose it, it's just how the file came out of a compiled C program), which in Notepad++ is shown as ``UCS-2 LE BOM``.

Further testing shows that any file that uses this or the corresponding big endian encoding is not shown in GitHub Desktop's ``Changes`` tab.
I believe those files should be displayed / their encodings should be supported.

Test file that will show this issue: [test.txt](https://github.com/desktop/desktop/files/3321613/test.txt)

**Comment by niik** [comment]
Hey @Morilli, thanks for opening this and thanks especially for providing a sample text file for us to reproduce with. After looking at this I believe this is the same issue as https://github.com/desktop/desktop/issues/3923 only manifested slightly differently. GitHub Desktop doesn't do any binary detection itself but instead relies on Git's internal binary detection and in this case the encoding was enough to trigger that. You can verify this yourself by manually running git diff on the file:

```
MacBook-Pro:desktop markus$ git add test.txt
MacBook-Pro:desktop markus$ git diff --cached
diff --git a/test.txt b/test.txt
new file mode 100644
index 000000000..2f8458609
Binary files /dev/null and b/test.txt differ
```

Given that I believe this to be a duplicate of #3923 I'm going to close this issue out and cross reference this issue with the reproduction file in it.

**Comment by Morilli** [comment]
So if GitHub Desktop is just using git's internal binary detection for such files, then the issue is solely on git's part, right? A fix or recognition of whether that's an actual issue or just a not supported thing must come from their side, correct?

**Comment by shiftkey** [comment]
@Morilli 

> So if GitHub Desktop is just using git's internal binary detection for such files, then the issue is solely on git's part, right? 

I believe so, but there's some extra details that matter here.

> A fix or recognition of whether that's an actual issue or just a not supported thing must come from their side, correct?

UTF-16 detection (UTF-16 being an extension of UCS-2) in Git out-of-the-box isn't great (I think this is slightly different to #3923), but there are some recent improvements in core Git that leverage the `.gitattributes` file along with the new `working-tree-encoding` attribute.

This is ideal for when you know an extension has a given encoding but it isn't detected by default, so I'm not sure if that's ideal for your situation. I'd have a read of the documentation and see if you can teach your Git repository to detect your file: https://git-scm.com/docs/gitattributes

**Comment by acrase** [comment]
I recently wasn't able to merge with master, so I deleted my branch and create a new branch. Since I did that, Desktop has a message that says "Binary file has changed. Open file in external program". What does this mean and how do I fix this? 

<img width="420" alt="Screen Shot 2020-08-14 at 9 32 54 AM" src="https://user-images.githubusercontent.com/62961335/90272278-04d3fa00-de12-11ea-96eb-df88b879bae7.png">

**Comment by tierninho** [comment]
@acrase Thanks for reaching out. Without seeing the exact file, the message in general means Desktop cannot render the file within the app, so use an external application to open the file to make any changes. An example would be a pdf file being opened by Adobe.

**Comment by acrase** [comment]
@tierninho thanks. You can go ahead and closed this. I was able to resolve the issue.

**Comment by mechPenSketch** [comment]
I have the same issue here, except I am sure I'm only using UTF-8 texts. On one of my private project, even when I have this on my .gitattribute:
```
* text=auto eol=lf
```
Github Desktop still says "The binary file has changed." for my text files. Strangely, this does not apply to my other projects, even when their .gitattribute has the same settings as my affected project for just have `text=auto`. Are there any other factors I can check for?

----

I thought it could be due to me setting the text settings in .gitattribute too late. However, when I check its history on the github website, I've found out that I have already set `* text=auto` on the initial commitment.

----

I have checked the line ending on my affected project's .gitattribute (Done on Notepad, goto View > Show Symbol > Show End Of Line) and found out it ends of CRLF and not LF. Probably to do with the previous version of Desktop, but it should be resolvable by making all the conversion accordingly.

But even with .gitattribute converted to LF, Github Desktop still has the "The binary file has changed: message displayed instead of listing text changes...does that mean CRLF wasn't the reason why I can't view text changes in Github Desktop?

Oh, I have checked the bottom-right corner of Notepad++ and it's indeed using UFT-8.

----

So I've found my some other projects being affected by it too. Public examples include:
* [My JRPG Fragment - Overworld (Dialogic)](https://github.com/mechPenSketch/Jrpg_Fragment-Overworld-Dialogic)
* [Battle System](https://github.com/mechPenSketch/Jrpg_Fragment-TurnBased_Combat)

On the website, going to a text file then History would display the text changes, but not in Github Desktop.

----

I have opened up Github desktop on another computer and curiously, my affected project has displayed my past text differences. However, I did not update it for a long time; it is only when I hit push origin did the differences become unviewable. This means there is some point among the commits that is responsible.
